### PR TITLE
feat: add max-header-length for header length check

### LINKED ISSUE #67: Commitlint fails after squash merge
**Describe the bug**
When a PR is created with a commit message of around 72 characters the commitlint action will pass. But when the PR is Squashed and Merged Github appends the PR number in the commit message. The appended commit message fails the commitlint after the PR is merged.
https://github.com/opensource-nepal/node-nepali-datetime/actions/runs/11974343317
Screenshot 2024-11-22 at 8 42 35 PM

**To Reproduce**
Pre-requisite: Github Action is set up on the target branch with `pull_request` and `push` events.
1. Create a PR with a commit message of 72 characters.
2. The commitlint will pass.
3. Squash and merge the PR.
4. Leave the commit message GitHub suggests (PR number appended to the commit message).
5. The commitlint fails after the PR is merged as the commit message length exceeds 72 characters.

**Expected behavior**
The commitlint should be smart enough to overcome this situation.
Possible Solution:
- Ignore the appended PR number; common formats `... (#PR)` and `... #PR`.

**Environment**
-   Python version: N/A
-   Package version: v1.3.0
-   Operating system: Github Actions
-   Any other relevant information about your environment.

### PR DESCRIPTION
**Pull request by aj3sh**
## Description
- add max-header-length for header length check
- disable default header length check
- update dependency actions version
- add release details on contributing file

## Related Issue

Fixes #67

## Type of Change

Please mark the appropriate option below to describe the type of change your pull request introduces:

-   [ ] Bug fix
-   [x] New feature
-   [x] Enhancement
-   [ ] Documentation update
-   [ ] Refactor
-   [ ] Other (please specify)

## Checklist

-   [x] My pull request has a clear title and description.
-   [x] I have used [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
        Examples: `"fix: Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
-   [x] I have added/updated the necessary documentation on `README.md`.
-   [x] I have added appropriate test cases (if applicable) to ensure the changes are functioning correctly.

## Additional Notes

[Add any additional notes or context that you think the reviewers should know about.]

By submitting this pull request, I confirm that I have read and complied with the contribution guidelines of this project.

### CODE REVIEW

**Review by sugat009 — COMMENTED** [review]
Supply chain security: please keep SHA pins for write-privileged actions. See inline comments.

File: .github/workflows/release-please.yml [hunk 1]
```diff
@@ -15,9 +15,9 @@ jobs:
     steps:
       - name: Release
         id: release
```
Comment (sugat009) [inline .github/workflows/release-please.yml:19]: **Suggestion: Keep SHA pin for this write-privileged action**

This action has `contents: write` and creates GitHub releases/tags. The current `main` branch correctly SHA-pins it:

```yaml
uses: googleapis/release-please-action@7987652d64b4581673a76e33ad5e98e3dd56832f # v4.1.3
```

Switching to a mutable tag (`@v4`) means anyone with write access to that repo can silently move the tag to point at different code — a known supply chain vector (ref: [March 2025 tj-actions hijack](https://blog.rafaelgss.dev/why-you-should-pin-actions-by-commit-hash)).

**Suggested change:**
```yaml
uses: googleapis/release-please-action@16a9c90856f42705d54a6fda1823352bdc62cf38 # v4
```

The `# v4` comment lets Dependabot auto-update the SHA when new versions release.

File: .github/workflows/release-please.yml [hunk 2]
```diff
@@ -49,4 +49,4 @@ jobs:
 
       - name: Publish package
         if: ${{ steps.release.outputs.release_created }}
-        uses: pypa/gh-action-pypi-publish@ed0c53931b1dc9bd32cbe73a98c7f6766f8a527e # v1.13.0
+        uses: pypa/gh-action-pypi-publish@v1.13.0
```
Comment (sugat009) [inline .github/workflows/release-please.yml]: **Suggestion: Keep SHA pin for PyPI publish action**

This action publishes packages to PyPI — the highest-privilege action in the repo. The current `main` branch correctly SHA-pins it:

```yaml
uses: pypa/gh-action-pypi-publish@ed0c53931b1dc9bd32cbe73a98c7f6766f8a527e # v1.13.0
```

`v1.13.0` resolves to the **exact same commit SHA** already pinned on `main`, so this line should simply remain unchanged.

For ongoing maintenance, consider adding a `.github/dependabot.yml` to auto-update SHA pins:
```yaml
version: 2
updates:
  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "weekly"
```

File: src/commitlint/linter/validators.py [hunk 3]
```diff
@@ -70,9 +71,15 @@ def validate(self) -> None:
         Returns:
             None
         """
+        max_header_length = config.max_header_length
+
+        if not max_header_length:
+            # skip header length check
+            return
+
```
Comment (sugat009) [inline src/commitlint/linter/validators.py:80]: **Bug: `not max_header_length` falsy check silently mishandles `0` and negative values**

`if not max_header_length:` uses Python truthiness — `0` is falsy, so it skips validation entirely. Negative values pass the guard and produce nonsensical behavior.

**Reproduction:**

```bash
# Bug 1: --max-header-length 0 silently passes (should reject or error)
$ commitlint --max-header-length 0 "feat: this is a very long commit message that should definitely fail any reasonable header length check but it wont"
Commit validation: successful!

# Bug 2: --max-header-length -5 rejects everything with a nonsensical message
$ commitlint --max-header-length -5 "feat: short"
✖ Found 1 error(s).
- Header length cannot exceed -5 characters.
```

**Suggested fix — two changes:**

1. Use explicit `None` check here instead of falsy:
```python
if max_header_length is None:
    return
```

2. Add input validation in `cli.py` to reject non-positive values at parse time:
```python
def positive_int(value):
    ivalue = int(value)
    if ivalue < 1:
        raise argparse.ArgumentTypeError(f"{value} is not a positive integer")
    return ivalue

parser.add_argument("--max-header-length", type=positive_int, ...)
```

The same falsy issue exists in `github_actions/action/run.py` where `if max_header_length:` should be `if max_header_length is not None:`.

**Review by sugat009 — COMMENTED** [review]
Bug: `--max-header-length` falsy check allows invalid values. See inline comment with reproduction steps.

File: src/commitlint/linter/validators.py [hunk 4]
```diff
@@ -70,9 +71,15 @@ def validate(self) -> None:
         Returns:
             None
         """
+        max_header_length = config.max_header_length
```
Comment (sugat009) [inline src/commitlint/linter/validators.py]: **Question: should validators read from the global config singleton, or receive parameters explicitly?**

With this change, `HeaderLengthValidator` now imports and reads `config.max_header_length` directly, making it the only validator with a hidden dependency on global state. The other validators are pure functions of `commit_message` alone, which keeps them easy to test and reason about.

A couple of things I'm wondering:

1. **Is this the pattern we want going forward?** If more options become configurable (e.g., `--max-body-line-length`, `--allowed-types`), each validator would import `config` and read its own values. The config singleton would grow from an infrastructure concern (verbose/quiet) into a bag of validator-specific settings, and every validator test would need to mock it.

2. **Would it be cleaner to pass `max_header_length` explicitly through the call chain?** For example, adding it as a parameter to `lint_commit_message()` (similar to how `skip_detail` and `strip_comments` are already passed), and then threading it through to the validator via a class attribute or `run_validators()`. That way `config` stays small and validators remain testable without patching globals.

3. **On the default value:** the existing `COMMIT_HEADER_MAX_LENGTH = 72` constant served as a sensible default. With this PR, header length checking is off unless explicitly opted in. Was that intentional, or would it make more sense to default to 72 and let `--max-header-length` override it? Keeping the default preserves backward compatibility for existing users.

Not blocking, just want to align on the direction before this sets the pattern for future configurable rules.

**Review by sugat009 — COMMENTED** [review]
Question on the architectural direction for configurable validator options. See inline comment.

File: github_actions/action/utils.py [hunk 5]
```diff
@@ -56,6 +56,27 @@ def get_boolean_input(key: str) -> bool:
...
+    """
+    val = get_input(key)
+
+    if not val:
+        return None
+
+    try:
+        return int(val)
+    except ValueError:
+        return None
```
Comment (sugat009) [inline github_actions/action/utils.py]: **issue (blocking):** `get_int_input` silently swallows invalid input

`get_boolean_input` raises `TypeError` on invalid values, but `get_int_input` returns `None` and silently disables the check. If a user writes `max_header_length: "abc"` in their workflow YAML, there is no warning, no error, and no indication the configuration was ignored.

```python
except ValueError:
    return None  # silently treated as "not configured"
```

This creates an inconsistency in the contract between the two input helpers and makes misconfiguration invisible to the user.

**Suggested fix:**

```python
except ValueError:
    raise ValueError(
        f"Input '{key}' must be a valid integer, got: {val!r}"
    )
```

File: src/commitlint/config.py [hunk 6]
```diff
@@ -14,6 +14,7 @@ class _CommitlintConfig:
 
     _verbose: bool = False
     _quiet: bool = False
```
Comment (sugat009) [inline src/commitlint/config.py:16]: **suggestion (non-blocking):** add a `@property`/setter for `max_header_length` to match the existing pattern

`verbose` and `quiet` both use `@property` with setters that enforce invariants. Adding `max_header_length` as a bare class attribute breaks that consistency and skips runtime validation entirely.

For example, `config.max_header_length = "not-an-int"` or `config.max_header_length = -5` would be accepted silently.

```python
_max_header_length: Optional[int] = None

@property
def max_header_length(self) -> Optional[int]:
    return self._max_header_length

@max_header_length.setter
def max_header_length(self, value: Optional[int]) -> None:
    if value is not None and value < 1:
        raise ValueError("max_header_length must be a positive integer")
    self._max_header_length = value
```

**nitpick (non-blocking):** `int | None` here vs `Optional["_CommitlintConfig"]` on line 13 is a style inconsistency within the same file. Both are valid for Python 3.10+, but picking one convention would keep things uniform.

File: src/commitlint/cli.py [hunk 7]
```diff
@@ -94,6 +94,14 @@ def get_args() -> argparse.Namespace:
         default=False,
     )
 
+    # --max-header-length : enables header length check (optional)
+    parser.add_argument(
+        "--max-header-length",
+        type=int,
+        help="Maximum header length to check",
```
Comment (sugat009) [inline src/commitlint/cli.py]: **nitpick (non-blocking):** help text could describe the default behavior and units

Currently:
```python
help="Maximum header length to check"
```

A user reading `--help` won't know what happens when this is omitted (no length check) or what "length" means (characters).

Suggestion:
```python
help="Maximum number of characters allowed in the commit header. "
     "If not specified, header length is not checked."
```

File: tests/test_cli.py [hunk 8]
```diff
@@ -369,19 +394,51 @@ def test__main__sets_config_for_verbose(
...
 
+    # main : max-header-length
+
     @patch(
         "commitlint.cli.get_args",
-        return_value=ArgsMock(file="path/to/non_existent_file.txt"),
+        return_value=ArgsMock(commit_message="feat: test commit", max_header_length=72),
     )
-    def test__main__with_missing_file(
+    def test__main__sets_config_for_max_header_length(
```
Comment (sugat009) [inline tests/test_cli.py]: **todo (non-blocking):** reset `config.max_header_length` in test teardown to prevent state leaks

The existing test fixtures reset `config.verbose` and `config.quiet` after each test class, but `max_header_length` is never cleaned up. Tests that call `main()` will set `config.max_header_length` on the singleton, and that value persists into subsequent tests.

This could cause flaky failures if test execution order changes or new tests are added that depend on the default `None` value.

**Review by sugat009 — COMMENTED** [review]
Additional review comments using [conventional comments](https://conventionalcomments.org/) labels.

**Review by sugat009 — CHANGES_REQUESTED** [review]

File: src/commitlint/linter/validators.py (hunk 4 above)
Comment (aj3sh) [reply src/commitlint/linter/validators.py]: There were two approaches: a global configuration or argument drilling. I was initially unsure which to choose, but argument drilling seems to be the better option.

I will update the code to pass the CLI arguments to the validators.

File: .github/workflows/release-please.yml (hunk 2 above)
Comment (aj3sh) [reply .github/workflows/release-please.yml]: Used commit hash, also added code comments regarding the highest-privilege action in the repo.

File: src/commitlint/linter/validators.py (hunk 3 above)
Comment (aj3sh) [reply src/commitlint/linter/validators.py:80]: Fixed as suggested.

File: src/commitlint/linter/validators.py (hunk 4 above)
Comment (aj3sh) [reply src/commitlint/linter/validators.py]: Created AppParams class for passing parameters nested. This introduced changes in lot of files 😀

File: github_actions/action/utils.py (hunk 5 above)
Comment (aj3sh) [reply github_actions/action/utils.py]: Fixed as suggested

File: src/commitlint/config.py (hunk 6 above)
Comment (aj3sh) [reply src/commitlint/config.py:16]: Removed max_header_length from config

File: src/commitlint/cli.py (hunk 7 above)
Comment (aj3sh) [reply src/commitlint/cli.py]: Updated.

File: tests/test_cli.py (hunk 8 above)
Comment (aj3sh) [reply tests/test_cli.py]: This test has been removed.

File: tests/test_linter/test__linter.py [hunk 9]
```diff
@@ -19,20 +19,22 @@ def fixture_data(request):
...
-    success, errors = lint_commit_message(commit_message, skip_detail=False)
+    success, errors = lint_commit_message(commit_message, AppParams(skip_detail=False))
     assert success == expected_success
     assert errors == expected_errors
 
 
 def test__lint_commit_message__skip_detail(fixture_data):
     commit_message, expected_success, _ = fixture_data
-    success, _ = lint_commit_message(commit_message, skip_detail=True)
+    success, _ = lint_commit_message(commit_message, AppParams(skip_detail=False))
```
Comment (sugat009) [inline tests/test_linter/test__linter.py]: issue (blocking): Copy-paste bug. This passes `skip_detail=False`, making it identical to `test_lint_commit_message` above. Should be `AppParams(skip_detail=True)` to exercise the `SimplePatternValidator` + `fail_fast` code path.

File: tests/test_github_actions/test_run/test_run_commitlint.py [hunk 10]
```diff
@@ -56,3 +56,21 @@ def test__run_commitlint__verbose(mock_check_output):
         text=True,
         stderr=subprocess.PIPE,
     )
+
+
+@patch(
+    "subprocess.check_output",
+    return_value="feat: valid commit message",
+)
+@patch.dict(os.environ, {**os.environ, "INPUT_MAX_HEADER_LENGTH": "72"})
```
Comment (sugat009) [inline tests/test_github_actions/test_run/test_run_commitlint.py:73]: issue (blocking): All four tests in this file fail when run in isolation:

```
pytest tests/test_github_actions/test_run/test_run_commitlint.py
# 4 failed — KeyError: 'INPUT_VERBOSE' or 'INPUT_MAX_HEADER_LENGTH'
```

This test patches `INPUT_MAX_HEADER_LENGTH` but not `INPUT_VERBOSE`. The three existing tests above patch `INPUT_VERBOSE` but not `INPUT_MAX_HEADER_LENGTH`. Since `run_commitlint()` now reads both env vars, each test needs both.

They only pass in the full suite because `set_github_env_vars()` from other test files leaks env vars without cleanup.

Fix: add both env vars to every `@patch.dict`:
```python
@patch.dict(os.environ, {**os.environ, "INPUT_VERBOSE": "false", "INPUT_MAX_HEADER_LENGTH": ""})
```

File: src/commitlint/linter/validators.py [hunk 11]
```diff
@@ -70,9 +72,15 @@ def validate(self) -> None:
         Returns:
             None
         """
+        max_header_length = self.params.max_header_length
+
+        if max_header_length is None:
```
Comment (sugat009) [inline src/commitlint/linter/validators.py:77]: question (blocking): This changes the default behavior from "always check 72 chars" to "never check unless opt-in." Users upgrading will silently lose header length enforcement.

The Conventional Commits spec doesn't mandate a header length, so opt-in is defensible, but this is a breaking change for anyone relying on the implicit 72-char guard.

Two options:
1. Keep as opt-in (current), but document as a breaking change and bump the major version.
2. Default to 72 (`--max-header-length` defaults to 72, add `--max-header-length 0` to disable).

Which is intended?

File: github_actions/action/utils.py [hunk 12]
```diff
@@ -56,6 +56,28 @@ def get_boolean_input(key: str) -> bool:
...
 
+def get_int_input(key: str) -> int | None:
+    """
+    Read the GitHub action integer input.
+
+    Args:
+        key: Input key.
+
+    Returns:
+        The integer value of the input. If not integer, returns None
```
Comment (sugat009) [inline github_actions/action/utils.py:67]: issue (non-blocking): Docstring says "If not integer, returns None" but the function raises `ValueError` for non-integer, non-empty strings (line 78). Only empty string returns `None`.

```python
Returns:
    The integer value of the input, or None if the value is an empty string.

Raises:
    ValueError: If the value is a non-empty, non-integer string.
```

File: github_actions/action/utils.py [hunk 13]
```diff
@@ -56,6 +56,28 @@ def get_boolean_input(key: str) -> bool:
...
+    """
+    val = get_input(key)
+
+    if val == "":
+        # GitHub Action passes empty data as a empty string ("")
+        return None
+
+    try:
+        return int(val)
```
Comment (sugat009) [inline github_actions/action/utils.py:79]: suggestion (non-blocking): The CLI validates `> 0` via `positive_int_type()`, but this function accepts negative and zero values. A user setting `max_header_length: 0` in their workflow YAML would get a confusing subprocess error instead of a clear message.

Consider adding a positive-integer check:
```python
result = int(val)
if result <= 0:
    raise ValueError(f"Input '{key}' must be a positive integer (> 0).")
return result
```

File: src/commitlint/cli.py [hunk 14]
```diff
@@ -230,15 +261,23 @@ def main() -> None:
...
+                    strip_comments=True,
+                ),
             )
         elif args.hash:
             console.verbose("commit message source: hash")
             commit_message = get_commit_message_of_hash(args.hash)
             _handle_commit_message(
-                commit_message, skip_detail=args.skip_detail, hide_input=args.hide_input
+                commit_message,
+                AppParams(
```
Comment (sugat009) [inline src/commitlint/cli.py:276]: suggestion (non-blocking): `AppParams` is constructed four times in `main()`, and three are identical (only the `--file` branch adds `strip_comments=True`). Adding a new field to `AppParams` requires four edits.

Consider:
```python
from dataclasses import replace

params = AppParams(
    skip_detail=args.skip_detail,
    hide_input=args.hide_input,
    max_header_length=args.max_header_length,
)
# file branch only:
_handle_commit_message(commit_message, replace(params, strip_comments=True))
# all other branches:
_handle_commit_message(commit_message, params)
```

File: tests/test_cli.py [hunk 15]
```diff
@@ -369,19 +406,38 @@ def test__main__sets_config_for_verbose(
         main()
         assert config.verbose is True
 
+    # main : max-header-length
```
Comment (sugat009) [inline tests/test_cli.py:409]: suggestion (non-blocking): `max_header_length` is tested only for the direct commit message path. The `--file`, `--hash`, and `--from-hash` branches also construct `AppParams` with `max_header_length` but have no corresponding tests. At minimum, a test for the `--file` path (which uniquely adds `strip_comments=True`) would catch branch-specific bugs.

File: src/commitlint/app_params.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    Represents runtime parameters that control linting behavior and output handling.
+
+    These parameters are typically derived from CLI arguments and define how
+    commit messages are validated and displayed.
+    """
+
+    # Skips the detailed error check (fails immediately without detail error message).
+    skip_detail: bool = False
+
+    # Hide input from stdout/stderr. Specially used by Github Actions.
```
Comment (sugat009) [inline src/commitlint/app_params.py]: nitpick (non-blocking): "Specially" → "Specifically" and "Github" → "GitHub" (capital H, matching the project's naming elsewhere).

File: src/commitlint/linter/validators.py [hunk 17]
```diff
@@ -30,8 +31,9 @@
 class CommitValidator(ABC):
     """Abstract Base validator for commit message."""
 
-    def __init__(self, commit_message: str) -> None:
+    def __init__(self, commit_message: str, params: AppParams) -> None:
         self._commit_message = commit_message
+        self.params = params
```
Comment (sugat009) [inline src/commitlint/linter/validators.py]: nitpick (non-blocking): `self.params` is public while `self._commit_message` and `self._errors` use the private convention. Consider `self._params` for consistency.

File: .github/workflows/release-please.yml [hunk 18]
```diff
@@ -1,23 +1,25 @@
...
       pull-requests: write
 
     steps:
       - name: Release
         id: release
-        uses: googleapis/release-please-action@7987652d64b4581673a76e33ad5e98e3dd56832f # v4.1.3
+        uses: googleapis/release-please-action@16a9c90856f42705d54a6fda1823352bdc62cf38 # v4.4.0
 
-      - uses: actions/checkout@v4
+      - uses: actions/checkout@v6
```
Comment (sugat009) [inline .github/workflows/release-please.yml]: suggestion (non-blocking): Lines 11-12 say "always pin actions to a specific commit hash," but this line and `actions/setup-python@v6` (line 37) use mutable version tags in a workflow with `id-token: write`, `contents: write`, and `pull-requests: write` permissions. Consider pinning to commit hashes, like you did for `release-please-action` and `gh-action-pypi-publish` in the same file.

**Review by sugat009 — CHANGES_REQUESTED** [review]
## Review: feat: add max-header-length

Good feature addition overall. The `AppParams` dataclass is well-proportioned, `positive_int_type` handles edge cases correctly, and the `validate_fns` rename removes ambiguity with the module name.

Two test bugs need fixing before merge, and one product decision needs resolution. See inline comments.

### Findings

| # | Label | File | Finding |
|---|-------|------|---------|
| 1 | issue (blocking) | `test__linter.py:29` | `skip_detail` test passes `False` instead of `True` (copy-paste bug) |
| 2 | issue (blocking) | `test_run_commitlint.py:65` | All 4 tests fail in isolation due to env var leakage |
| 3 | question (blocking) | `validators.py:77` | Breaking change: header length check is now opt-in (was always 72) |
| 4 | issue | `utils.py:67` | Docstring claims "returns None" but raises `ValueError` |
| 5 | suggestion | `utils.py:76` | Add positive-int validation to match CLI path |
| 6 | suggestion | `cli.py:276` | DRY: extract `AppParams` construction with `dataclasses.replace` |
| 7 | suggestion | `test_cli.py:409` | Missing test coverage for `--file`/`--hash`/`--from-hash` + max-header-length |
| 8 | nitpick | `app_params.py:18` | Typo: "Specially" and "Github" |
| 9 | nitpick | `validators.py:36` | `self.params` public vs `self._commit_message` private |
| 10 | suggestion | `release-please.yml:22` | Pin actions to commit hashes in privileged workflow |

File: tests/test_linter/test__linter.py (hunk 9 above)
Comment (aj3sh) [reply tests/test_linter/test__linter.py]: Updated

File: tests/test_github_actions/test_run/test_run_commitlint.py (hunk 10 above)
Comment (aj3sh) [reply tests/test_github_actions/test_run/test_run_commitlint.py:73]: Updated

File: github_actions/action/utils.py (hunk 12 above)
Comment (aj3sh) [reply github_actions/action/utils.py:67]: Updated

File: github_actions/action/utils.py (hunk 13 above)
Comment (aj3sh) [reply github_actions/action/utils.py:79]: Updated

File: tests/test_cli.py (hunk 15 above)
Comment (aj3sh) [reply tests/test_cli.py:409]: Updated. Added test for all branches.

File: src/commitlint/app_params.py (hunk 16 above)
Comment (aj3sh) [reply src/commitlint/app_params.py]: Updated

File: src/commitlint/linter/validators.py (hunk 17 above)
Comment (aj3sh) [reply src/commitlint/linter/validators.py]: Updated

File: .github/workflows/release-please.yml (hunk 18 above)
Comment (aj3sh) [reply .github/workflows/release-please.yml]: actions/* are maintained by Github Actions, which are trusted. However, pining the commit hash is the best. Therefore updated.

File: src/commitlint/cli.py (hunk 14 above)
Comment (aj3sh) [reply src/commitlint/cli.py:276]: `AppParams` are only constructed at once, because of the branching. The current approach doesn't have any performance or readability issue. Using `replace` seems to make more confusion.

> Adding a new field to `AppParams` requires four edits.

Let's consider this when the new field is added to the `AppParams`.

File: src/commitlint/linter/validators.py (hunk 11 above)
Comment (aj3sh) [reply src/commitlint/linter/validators.py:77]: Since we are bumbing the major version for Node 24 (#89), this will also be treated as breaking change.

Will update the commit once the #89 is merged
